feature: implement popup windows

mattkae opened the pull request:
This relies on the engine built from this branch: https://github.com/flutter/engine/pull/56094

Design doc: https://docs.google.com/document/d/1eQG-IS7r4_S9_h50MY_hSGwUVtgSTRSLzu7MLzPNd2Y/edit?tab=t.0

This PR is a part of a group of PRs. Going from first to last, these must be merged in the following order:

- #157515
- #157517
- #157518
- #157519 
- #157520
- #157521 
- #157522 
- #157523 
- #157524 
- #157525

Features:
- Implement `createPopup`
- Add popup windows to the `multi_window_ref_app`


https://github.com/user-attachments/assets/5d9c3603-6710-41a6-b351-50c509ae7e44


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

mattkae committed:
feature: implement createRegularWindow and add it to the reference app

mattkae committed:
feature: implement createRegularWindow and add it to the reference app

mattkae committed:
feature: implement popup windows

mattkae committed:
wip: refactor of the foundational code

mattkae committed:
wip: refactor of the foundational code

mattkae committed:
wip: refactored the mutlti window reference app to use new API

mattkae committed:
wip: refactored the mutlti window reference app to use new API

mattkae committed:
wip: able to create windows in the reference application again

mattkae committed:
wip: able to create windows in the reference application again

mattkae committed:
Selection and rerenders are functioning appropriately

mattkae committed:
Selection and rerenders are functioning appropriately

mattkae committed:
Able to remove an application, albeit with some bugs + refactored how settings are donw

mattkae committed:
Able to remove an application, albeit with some bugs + refactored how settings are donw

mattkae committed:
Able to add and remove windows

mattkae committed:
Able to add and remove windows

mattkae committed:
All windows are now showing in the list

mattkae committed:
All windows are now showing in the list

mattkae committed:
Added the ability to listen for errors and respond to them accordingly

mattkae committed:
Added the ability to listen for errors and respond to them accordingly

mattkae committed:
RegularWindowMetadata and WindowMetadata are useless

mattkae committed:
RegularWindowMetadata and WindowMetadata are useless

mattkae committed:
The future should be set in setState

mattkae committed:
The future should be set in setState

mattkae committed:
Properly handling the event where we need to destroy the widget during a dispose

mattkae committed:
Properly handling the event where we need to destroy the widget during a dispose

mattkae committed:
Introducing new types + refactoring a GenericWindow widget out

mattkae committed:
Fix issue where setState returned a future

mattkae committed:
Fix issue where setState returned a future

mattkae committed:
setState is ssynchronous, whoops!

mattkae committed:
setState is ssynchronous, whoops!

mattkae committed:
Refactored window.dart to be more concise for the initial PR

mattkae committed:
Refactored window.dart to be more concise for the initial PR

mattkae committed:
Fixed issue where we might accidentally dispose of a window twice

mattkae committed:
Fixed issue where we might accidentally dispose of a window twice

mattkae committed:
Making the WindowMetadata public and only refresing the relevant fields when settings change

mattkae committed:
Making the WindowMetadata public and only refresing the relevant fields when settings change

mattkae committed:
Wip

mattkae committed:
It does not make sense for the modify method to be an override since different window types will have different values

mattkae committed:
It does not make sense for the modify method to be an override since different window types will have different values

mattkae committed:
Merge branch 'foundation-refactor' of github.com:canonical/flutter into foundation-refactor

mattkae committed:
Window settings dialog for anchor info and popup info

mattkae committed:
Popups now have the selected parent

mattkae committed:
bugfix: able to select the initial wndow in the table

mattkae committed:
bugfix: able to select the initial wndow in the table

mattkae committed:
Reimplement the rotating cube in the demo app

mattkae committed:
Reimplement the rotating cube in the demo app

mattkae committed:
Popups can be opened up now, except on the main window

mattkae committed:
Can render child windows from the primary window

mattkae committed:
PR feedback

mattkae committed:
PR feedback

mattkae committed:
Rendering correct text

mattkae committed:
Rendering correct text

mattkae committed:
Merge branch 'foundation-refactor' into popup-windows-refactor

mattkae committed:
Merge remote-tracking branch 'origin/master' into foundation

mattkae committed:
Merge branch 'foundation' into foundation-refactor

mattkae committed:
Reverting back to the old runner code

mattkae committed:
Reverting back to the old runner code

mattkae committed:
Merge branch 'master' into foundation

mattkae committed:
Merge branch 'foundation' into foundation-refactor

mattkae committed:
Merge pull request #11 from canonical/foundation-refactor

mattkae committed:
Update runner code to work with the latest & greatest

mattkae committed:
Update runner code to work with the latest & greatest

mattkae committed:
Merge branch 'foundation-refactor' into popup-windows-refactor

mattkae committed:
Updating CmakeLists to accomodate latest

mattkae committed:
Massive simplification of the the settings dialog

mattkae committed:
Custom positioner dialogs and settings

mattkae committed:
Properly applying the anchor rect

mattkae committed:
Able to scroll the regular window

mattkae committed:
Merge branch 'foundation' into popup-windows-refactor

hbatagelo committed:
Add command-line option for multi-window support and refactor runner code

hbatagelo committed:
Add command-line option for multi-window support and refactor runner code

mattkae committed:
Merge pull request #12 from canonical/add-command-line-option

Add `--enable-multi-window` option and refactor runner code

mattkae committed:
Merge branch 'foundation' into popup-windows-refactor

mattkae committed:
Initial PR feedback

mattkae committed:
Popup now scrolls and expands appropriately

mattkae committed:
For popups, we now clamp the anchor rectangle to the parent window size

mattkae committed:
Update examples/multi_window_ref_app/windows/flutter/CMakeLists.txt

Co-authored-by: Harlen Batagelo <hbatagelo@gmail.com>

mattkae committed:
minor: fix typo in comment

mattkae committed:
Merge branch 'popup-windows-refactor' of github.com:canonical/flutter into popup-windows-refactor

mattkae committed:
Updating popup window documentation comments

mattkae committed:
Reset all files to foundation-refactor

mattkae committed:
Merge branch 'foundation' into popup-windows

mattkae committed:
Merge branch 'popup-windows-refactor' into popup-windows

mattkae committed:
feature: adding full client-side tests for regular window creation  (#14)

feature: adding full client-side tests for regular window creation

mattkae committed:
Merge

mattkae committed:
Fix merge

mattkae committed:
Fix merge

mattkae committed:
Remove window.dart tests

jtmcdole commented:
Can you close and reopen this PR on upstream/master?  It is dated from pre-monorepo and will continue failing tests.

hbatagelo committed:
Add foundational support for Windows multi-window in the engine (#15)

hbatagelo committed:
Add support for multi-window popups on Windows (#16)

mattkae committed:
bugfix: super.dispose() was supposed to come first

mattkae committed:
Running dart format

mattkae committed:
refactor: moving the event listeners to a global instead of an inherited widget

mattkae committed:
testing: tests for popup-windows (#18)

mattkae committed:
No longer attempting to erroneously use the implicitView

hbatagelo committed:
Refactor method channel for regular windows

hbatagelo committed:
Review feedback + align code style with guidelines

hbatagelo committed:
Add FML as a dependency to common_cpp_core; fix incorrect include

hbatagelo committed:
Improve comments and code style consistency

mattkae committed:
Merge pull request #19 from canonical/refactor-method-channel

Refactor method channel for regular windows

mattkae committed:
Refactor regular windows (#20)

* refactor: the RegularWindowController now holds all initial values and is solely responsible for creating and destroying a window

* Updated multi window reference app and window tests

mattkae committed:
refactor: remove references to WindowingApp from the window.dart API (#21)

* refactor: remove references to WindowingApp from the window.dart API

* bugfix: notify listeners when values change on a window controller

mattkae committed:
refactor: raw create/destroy methods are now hidden + renamed WindowContext to WindowControllerContext and made it provide the controller (#22)

mattkae committed:
Merge branch 'foundation' into popup-windows-merge

mattkae committed:
feature: implemented modifyRegular + fixed a bug where a widget tree was mandatory on accident

mattkae committed:
feature: tidy up all of the documentation around regular windows + simplify parts of the API (#25)

hbatagelo committed:
Resolve remaining conflicts and update wrt the latest refactoring

hbatagelo committed:
Formatting nits

mattkae committed:
Merge pull request #23 from canonical/popup-windows-merge

Popup windows merge

mattkae committed:
Merge wip

mattkae committed:
refactor: distinguish a WindowCreationResult from a particular RegularWindowCreationResult

mattkae committed:
minor: fix spelling mistake

mattkae committed:
git is confused and not recognizing my pushes...?

mattkae committed:
Merge window.dart, woohoo!

mattkae committed:
Merging

mattkae committed:
merge: fix tests

mattkae committed:
Merge pull request #27 from canonical/popup-windows-merge-two

Popup windows merge two

mattkae committed:
Using ChildWindowRenderer for the main window + disallowing parentless popups

mattkae committed:
Merge branch 'popup-windows-merge' of github.com:canonical/flutter into popup-windows-merge

hbatagelo committed:
Add support for modifyRegular in the embedder and update examples/tests

hbatagelo committed:
Better ModifyRegularWindow test

hbatagelo committed:
Add inequality operator for flutter::Size

hbatagelo committed:
Add SetState and remove unused GetState

hbatagelo committed:
Log error messages during window creation, window change, and destruction

hbatagelo committed:
Split up tests for modifying the window and add a test for changing the window state

mattkae committed:
Merge pull request #24 from canonical/regular-windows-modify

feature: implemented modifyRegular + fixed a bug where a widget tree was mandatory on accident

mattkae committed:
Merge pull request #28 from canonical/example/adding-edit

example: adding a modificatione example to the multi window reference app

mattkae committed:
Merge remote-tracking branch 'upstream/master' into foundation

mattkae committed:
example: adding a modificatione example to the multi window reference app

hbatagelo committed:
Use kBadArgumentsError

mattkae committed:
Merge branch 'foundation' into popup-windows-merge

mattkae committed:
Fix merge

mattkae committed:
Apply et format

mattkae committed:
feedback: WindowControllerContext.of now returns the WindowController instead of the context

mattkae committed:
feature: add MaterialApp.useWindowingApi flag

hbatagelo committed:
Merge pull request #31 from canonical/material_app_windowing_flag

feature: add MaterialApp.useWindowingApi flag

hbatagelo committed:
Add support for windows created in the runner

hbatagelo committed:
Use SetProp/GetProp for all windows + tweak comments for consistency

hbatagelo committed:
Fix checks and assertions; early forward to lifecycle manager

hbatagelo committed:
Ensure the host window controller is destroyed before the lifecycle manager

hbatagelo committed:
Merge pull request #32 from canonical/use-runner-windows

Add support for windows created in the runner

hbatagelo committed:
Merge remote-tracking branch 'upstream/master' into foundation

mattkae committed:
Merge branch 'master' into foundation

hbatagelo committed:
Merge branch 'foundation' into popup-windows

mattkae committed:
Merge branch 'master' into foundation

mattkae committed:
Revert all API changes to the framework

mattkae committed:
Revert "Revert all API changes to the framework"

This reverts commit 1070b86d78bb57d25fc1a5b039274b34bd764902.

hbatagelo committed:
Merge branch 'master' into foundation

hbatagelo committed:
Simplify focusing the root view when the top-level window is activated

hbatagelo committed:
Update reference app to ensure root views of regular windows remain in the same view collection

mattkae committed:
Simplify focusing the root view when the top-level window is activated (#37)

This simplifies the logic for focusing the root view when the top-level
window is activated. It removes the handling of `WM_MOUSEACTIVATE`, as
`WM_ACTIVATE` already accounts for activations via mouse clicks.
Additionally, it ensures that the view is not focused when the top-level
window is being deactivated.

mattkae committed:
Update reference app to ensure root views of regular windows remain in the same view collection (#38)

This PR updates the reference app so that the root view of a regular
window stays in the topmost view collection instead of being anchored to
the view of the main window. This ensures that the `FocusTraversalGroup`
of the root view is a child of the root focus scope rather than a nested
scope.

Why does this matter? Because it fixes an issue where focus sometimes
switches back and forth indefinitely between two regular windows.
Specifically, the issue occurs when switching focus from a root view A
to a root view B while B's focus scope is nested in A's focus scope.
When the user activates the top-level window with root view B and the
focus goes to B, A's `View::_scopeFocusChangeListener` detects that the
focused scope does not match the newly focused view (B) and requests a
focus change back to A. However, when A regains focus, B's
`View::_scopeFocusChangeListener` then notices that B is not focused
while its scope node is (since it's inside A's scope) and requests a
focus change to B, restarting the loop.

mattkae committed:
Merge branch 'master' into foundation

mattkae committed:
bugfix: get rid of lifecycle state

mattkae committed:
Merge branch 'foundation' of github.com:canonical/flutter into foundation

hbatagelo committed:
Remove unused key and add offset type for relative_position

mattkae committed:
Merge branch 'foundation' into foundation-plus-framework

hbatagelo committed:
Merge branch 'foundation-plus-framework' into popup-windows

hbatagelo committed:
Merge branch 'master' into foundation

hbatagelo committed:
Merge branch 'foundation' into foundation-plus-framework

hbatagelo committed:
Merge branch 'foundation-plus-framework' into popup-windows

hbatagelo committed:
Merge branch 'master' into foundation

hbatagelo committed:
Merge branch 'foundation' into popup-windows

mattkae commented:
The implementation of multi-window in Flutter has drifted a lot since this PR was opened. Please see https://github.com/flutter/flutter/pull/168728 for the latest implementation. We're currently focusing on an FFI-based approach in contrast to the async method channels approach that we implemented here.
